gnc/sub8_slam/src/slam.cc

```diff
@@ -0,0 +1,254 @@
+      prev_frame = new_frame.clone();
+      continue;
+    }
+
+    ////// Track known points and dismiss outliers
+    slam::StatusVector status;
+    slam::optical_flow(prev_frame, new_frame, prev_feature_locations, new_feature_locations,
+                       status);
+    // Determine the IDs of the preserved points
+    // TODO: Filter and which_pts in the same function
+    new_feature_ids = slam::which_points(status, prev_feature_ids);
+    new_feature_locations = slam::filter(status, new_feature_locations);
+    prev_feature_locations = slam::filter(status, prev_feature_locations);
+
+    ////// Compute the fundametnal matrix and dismiss outliers
+    slam::StatusVector inliers;
+    cv::Mat F;
+    F = slam::estimate_fundamental_matrix(prev_feature_locations, new_feature_locations, inliers);
+    new_feature_ids = slam::which_points(inliers, new_feature_ids);
+    // Filter by inlier mask
+    new_feature_locations = slam::filter(inliers, new_feature_locations);
+    prev_feature_locations = slam::filter(inliers, prev_feature_locations);
+
+    ////// Estimate motin from the fundamental matrix
+    slam::Pose pose_F;
+    pose_F = slam::estimate_motion_fundamental_matrix(prev_feature_locations, new_feature_locations,
+                                                      F, intrinsics);
+
+    ////// Triangulate points and dismiss the guess based on magnitude of reprojection error
+
+    slam::Point3Vector triangulation_F;
+    slam::triangulate(prev_pose, pose_F, intrinsics, prev_feature_locations, new_feature_locations,
+                      triangulation_F);
+
+    // Todo: use PNP pose
+    double error_amt = slam::average_reprojection_error(triangulation_F, new_feature_locations,
+                                                        pose_F, intrinsics);
+    // std::cout << "Error: " << error_amt << std::endl;
+
+    ////// Initialize a slam frame based on these shenanigans
+    if (error_amt < 100.0) {
+      // initialize map only if we have little reproj error
+      if (map.size() == 0) {
+        map = triangulation_F;
+      }
+
+      slam::Point3Vector map_visible;
+      // Determine which points in the map are currently visible
+      map_visible = slam::get_points(new_feature_ids, map);
+      slam::Pose pose_pnp;
+      pose_pnp = slam::estimate_motion_pnp(map_visible, new_feature_locations, intrinsics);
+
+      double pnp_error_amt = slam::average_reprojection_error(map_visible, new_feature_locations,
+                                                              pose_pnp, intrinsics);
+
+      std::cout << "PNP Error: " << pnp_error_amt << std::endl;
+      slam::draw_reprojection(render_frame, map_visible, pose_pnp, intrinsics);
+
+      // Magic number -- TODO why is this so large compared to 100.0?
+      if (pnp_error_amt < 2500.0) {
+        slam::Point3 pt(pose_pnp.translation);
+        camera_locations.push_back(pt);
+
+        slam::Frame key_frame(pose_pnp, new_feature_ids, new_feature_locations);
```

Comment:
The idea here (in the next iteration) is to initialize a new keyframe (and associated map section) when the reprojection error becomes too high. That is, we'd backtrack `k` frames from the time when pnp_error exceeded some threshold.

The blocking issue for this is that reprojections from PnP pose solutions are flawed somehow. I don't think switching to sparse model-based image alignment on SE(3) is going to fix it yet.